### PR TITLE
bugfix: Try to fix Mill flaky test

### PR DESCRIPTION
**Pull request by tgodzik**
My guess is that the test failed because it takes a while for the build server to connect. Let's see if this helps.

### CODE REVIEW

File: tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala [hunk 1]
```diff
@@ -40,6 +43,8 @@ class MillServerCodeLensSuite
       _ <- server.initialize()
       _ <- server.initialized()
       _ <- server.executeCommand(ServerCommands.GenerateBspConfig)
+      _ <- server.server.buildServerPromise.future
+        .withTimeout(2, TimeUnit.MINUTES)
```
Comment (kasiaMarek) [inline tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala:47]: Doesn't `didOpen` wait for the `buildServerPromise` anyway? I though it's more of a when the bsp compiles and publishes things not the connection itself.

Comment (tgodzik) [reply tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala]: Might be, but I can't figure out where does the test actually hang :/

Comment (kasiaMarek) [reply tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala]: Maybe add printing of the bsp trace on error. That could provide some insight.

File: tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala [hunk 2]
```diff
@@ -14,8 +18,15 @@ class MillServerCodeLensSuite
 
   override def munitTimeout: Duration = Duration("4min")
 
+  override def afterEach(context: AfterEach): Unit = {
+    super.afterEach(context)
+    scribe.info(workspace.resolve(".metals/bsp.trace.json").text)
```
Comment (kasiaMarek) [inline tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala:23]: Can you print it only if the test fails?

Comment (tgodzik) [reply tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala:23]: I think it's better not to print at all as writing to bsp.trace.json might slow down tests. I will remove it afterwards.

Comment (tgodzik) [reply tests/slow/src/test/scala/tests/mill/MillServerCodeLensSuite.scala:23]: Ach and it's typical, when adding logs it works :sweat_smile:

**Comment by tgodzik** [comment]
I think I finally found the reason the tests hung and weren't retried. Looks like Mill will sometimes throw an exception which is forwarded to Metals and causes tests to hange, this should make the tests stop hanging and make retries work.

**Comment by tgodzik** [comment]
> Still seems a mill test failed do you know why?

Damn, no idea. I need to figure out where it hangs, but I don't see any other possibility. I will add a manual timeout maybe.

**Comment by tgodzik** [comment]
Looks like finally the approach worked and I run the mill tests 3 times and tests didn't fail. What do you think @kasiaMarek ?

**Comment by kasiaMarek** [comment]
> Looks like finally the approach worked and I run the mill tests 3 times and tests didn't fail. What do you think @kasiaMarek ?

If it's mill that throws for undisclosed reasons, then it's fine I think. However, I would be great to explain it in a comment and take some steps to try and resolve the underlying problem. Maybe create an issue in `mill`?

**Comment by tgodzik** [comment]
> > Looks like finally the approach worked and I run the mill tests 3 times and tests didn't fail. What do you think @kasiaMarek ?
> 
> If it's mill that throws for undisclosed reasons, then it's fine I think. However, I would be great to explain it in a comment and take some steps to try and resolve the underlying problem. Maybe create an issue in `mill`?

I opened up an issue in mill https://github.com/com-lihaoyi/mill/issues/2826 I am not 100% that's the reason, but this is the only stack I've seen recently. I added a comment around that test suite, so people might see it's problematic.

File: tests/unit/src/main/scala/tests/TestingServer.scala [hunk 3]
```diff
@@ -1034,6 +1034,7 @@ final case class TestingServer(
       filename: String,
       maxRetries: Int = 4,
   ): Future[List[l.CodeLens]] = {
+    Debug.printEnclosing(filename)
```
Comment (kasiaMarek) [inline tests/unit/src/main/scala/tests/TestingServer.scala:1037]: Can you explain to me what does this thing do?

**Review by kasiaMarek — APPROVED** [review]
lgtm

File: tests/unit/src/main/scala/tests/TestingServer.scala (hunk 3 above)
Comment (tgodzik) [reply tests/unit/src/main/scala/tests/TestingServer.scala:1037]: just prints that this method has been invoked together with the filename. 

This is where things like `2023.10.06 10:30:46 INFO  tests.BaseLspSuite#initialize ` come from.

**Pull request merged**